yt_dlp/extractor/duoplay.py

```diff
@@ -0,0 +1,81 @@
+from .common import InfoExtractor
+from ..utils import (
+    extract_attributes,
+    get_element_text_and_html_by_tag,
+    int_or_none,
+    join_nonempty,
+    str_or_none,
+    try_call,
+    unified_timestamp,
+)
+from ..utils.traversal import traverse_obj
+
+
+class DuoplayIE(InfoExtractor):
+    _VALID_URL = r'https://duoplay\.ee/(?P<id>\d+)/[\w-]+/?(?:\?(?:[^#]+&)?ep=(?P<ep>\d+))?'
+    _TESTS = [{
+        'note': 'Siberi võmm S02E12',
+        'url': 'https://duoplay.ee/4312/siberi-vomm?ep=24',
+        'md5': '1ff59d535310ac9c5cf5f287d8f91b2d',
+        'info_dict': {
+            'id': '4312_24',
+            'ext': 'mp4',
+            'title': 'Operatsioon "Öö"',
+            'thumbnail': r're:https://.+\.jpg(?:\?c=\d+)?$',
+            'description': 'md5:8ef98f38569d6b8b78f3d350ccc6ade8',
+            'upload_date': '20170523',
+            'timestamp': 1495567800,
+            'series': 'Siberi võmm',
+            'series_id': '4312',
+            'season': 'Season 2',
+            'season_number': 2,
+            'episode': 'Operatsioon "Öö"',
+            'episode_number': 12,
+        },
+    }, {
+        'note': 'Empty title',
+        'url': 'https://duoplay.ee/17/uhikarotid?ep=14',
+        'md5': '6aca68be71112314738dd17cced7f8bf',
+        'info_dict': {
+            'id': '17_14',
+            'ext': 'mp4',
+            'title': 'Episode 14',
+            'thumbnail': r're:https://.+\.jpg(?:\?c=\d+)?$',
+            'description': 'md5:4719b418e058c209def41d48b601276e',
+            'upload_date': '20100916',
+            'timestamp': 1284661800,
+            'series': 'Ühikarotid',
+            'series_id': '17',
+            'season': 'Season 2',
+            'season_number': 2,
+            'episode': 'Episode 14',
+            'episode_number': 14,
+        },
+    }]
+
+    def _real_extract(self, url):
+        telecast_id, episode = self._match_valid_url(url).groups()
+        video_id = join_nonempty(telecast_id, episode, delim='_')
+        webpage = self._download_webpage(url, video_id)
+        video_player = try_call(lambda: extract_attributes(
+            get_element_text_and_html_by_tag('video-player', webpage)[1]))
+        if not video_player or not video_player.get('manifest-url'):
+            self.raise_no_formats('No video found', expected=True)
+
+        episode_attr = self._parse_json(video_player.get(':episode') or '', video_id, fatal=False) or {}
+
+        return {
+            'id': video_id,
+            'formats': self._extract_m3u8_formats(video_player['manifest-url'], video_id, 'mp4'),
+            **traverse_obj(episode_attr, {
+                'title': (None, ('subtitle', ('episode_id', {lambda x: f'Episode {x}'}))),
```

Comment:
This now brings up the problem of the title. This is the best I came up with, tell me if you think this makes sense:
For movies, always use the title. For shows/series, first try the subtitle, then try to name it after the episode number and if all else fails also use the title.
```suggestion
            'title': episode_attr.get('title') if episode_attr.get('category') == 'movies' else (
                traverse_obj(episode_attr, 'subtitle', ('episode_nr', {lambda x: f'Episode {x}' if x else None}), 'title')),
            'formats': self._extract_m3u8_formats(video_player['manifest-url'], video_id, 'mp4'),
            **traverse_obj(episode_attr, {
```
This means the tests will need to be updated. But I think it would be unexpected to have for example `https://duoplay.ee/17/uhikarotid?ep=14` be named `Episode 14` even though it is the 2nd episode of the 2nd season.

Reply:
I thought of movies in a separate pr with same logic, and my version was more verbose. 

sadly for the yhikarotid, you have to accept the absolute number scheme, or rename items manually. maybe if show has only absolute numbers, then omit the season 01 from it's name?

```diff
diff --git a/yt_dlp/extractor/duoplay.py b/yt_dlp/extractor/duoplay.py
index f12c91012..d2dbc27c4 100644
--- a/yt_dlp/extractor/duoplay.py
+++ b/yt_dlp/extractor/duoplay.py
@@ -53,8 +53,15 @@ class DuoplayIE(InfoExtractor):
         },
     }]
 
+    @staticmethod
+    def is_movie(attr):
+        # return attr.get('episode_id', None) == 0
+        return attr.get('category', None) == 'movies'
+
     def _real_extract(self, url):
         telecast_id, episode = self._match_valid_url(url).groups()
         video_id = join_nonempty(telecast_id, episode, delim='_')
         webpage = self._download_webpage(url, video_id)
         video_player = try_call(lambda: extract_attributes(
@@ -63,14 +70,28 @@ def _real_extract(self, url):
             self.raise_no_formats('No video found', expected=True)
 
         episode_attr = self._parse_json(video_player.get(':episode') or '', video_id, fatal=False) or {}
 
-        return {
+        data = {
             'id': video_id,
             'formats': self._extract_m3u8_formats(video_player['manifest-url'], video_id, 'mp4'),
+            'description': 'synopsis',
+            'thumbnail': traverse_obj(episode_attr, ('images', 'original')),
+        }
+
+        if self.is_movie(episode_attr):
+            return {
+                **data,
+                'title': f"{traverse_obj(episode_attr, 'title')} ({traverse_obj(episode_attr, 'year')})",
+            }
+
+        return {
+            **data,
             **traverse_obj(episode_attr, {
                 'title': (None, ('subtitle', ('episode_id', {lambda x: f'Episode {x}'}))),
-                'description': 'synopsis',
-                'thumbnail': ('images', 'original'),
                 'timestamp': ('airtime', {lambda x: unified_timestamp(x + ' +0200')}),
                 'series': 'title',
                 'series_id': ('telecast_id', {str_or_none}),
```

Reply:
> maybe if show has only absolute numbers, then omit the season 01 from it's name?

That could work.
If no season number present ⇒ absolute value. If season number present ⇒ relative value.
I'd be curious to hear what one of the maintainers thinks.

As for your diff, I see that you are excluding season and episode data on movies, which makes sense. I'll make a better suggestion later working that in.

Reply:
I added test data for a movie. and yes, the series data doesn't make sense for movie. also id like to put `year` to movie title. can you help with that too?

Reply:
also. the movie will expire in 4 days, not sure if they remove the URL and data if that expires.

```json
  "_visible_until": {
    "diff": 4,
    "unit": "day"
  }
```

Reply:
and, for the show with absolute numbering, if they also lack subtitle, the episode downloaded will be named with just show title... well. ok final filename contains also the id, so it's "Ühikarotid [17_14].mp4" now, but could be "Ühikarotid - Episode 14 [17_14].mp4"

i.e. this test URL:
- https://duoplay.ee/17/uhikarotid?ep=14

Reply:
lol. the test movie expires in 3 hours: "Vaadatav 3 tundi". should set that test data to skip?

Reply:
so, found one "movie" (in translation it says "special episode") without an expiry date:
- https://duoplay.ee/8279/mulle-eestist-ei-piisa

and that's reflected in the metadata too: `"category": "shows"`

Reply:
Added test data for movie that doesn't expire: 29ac66b42239eb57c521ecfd1a4beddc86531e31

what to do with the item that will expire (of course have to wait few hours to see how it appears on the site)

Reply:
so. the movie has expired and returns 404. Should the test be dropped or use some skip attribute?

```
✖ curl -i https://duoplay.ee/4325/naljamangud
HTTP/2 404 
content-language: et-ee
x-frame-options: SAMEORIGIN
content-type: text/html; charset=utf-8
date: Fri, 17 Nov 2023 08:57:30 GMT
via: 1.1 varnish (Varnish/6.0)
vary: Accept-Encoding, User-Agent
cache-control: no-cache
age: 0
content-length: 33749

[...]
```

Reply:
I will add it in the next cleanup PR/commit. Thanks for the reminder